### PR TITLE
Akka basics.

### PR DESCRIPTION
**Pull request by felipefzdz**

### CODE REVIEW

File: _posts/2016-04-30-akka-basics.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      text: Akka_20120331 by Helena Whitlock, SGU
+      href: https://www.flickr.com/photos/geologicalsurveyofsweden/8703286181
+tags:
+- akka
+- OOP
+- functional programming
+
+---
+
+As [the previous post](http://codurance.com/2016/04/28/async-systems-with-sync-clients) explained our system is based on Akka. Before going into more detail about our solution, I'd like to explain what is Akka about and why is so awesome.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: Might be better if you close the sentence as:
what Akka is about and why it is so awesome.

File: _posts/2016-04-30-akka-basics.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+Akka is a toolkit and not a framework, you can simply use the bits that you need for your service. In this series we'll focus on the core features and we won't go through Akka Stream, Akka-Http or Akka Cluster (since they're out of scope, not because they're not fantastic).
+
+Akka provides a different abstraction to deal with concurrency, paralellism and fault-tolerance. That abstraction is called Actor Model. If you have experience with old Java ecosystem you will know how hard is writing safe and correct multithreaded code. Providing human-friendly abstractions like Actors it's a revolution similar to what Java did with memory management.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: ...how hard it is to write safe and correct... over here refer to your previous posts on threads.

Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: Reading this paragraph I feel you can still say something on what high level problem Akka or Reactive systems solve - maybe even refer to the architectural pattern its related to - Event Driven Architecture (https://en.wikipedia.org/wiki/Event-driven_architecture) if that comes across applicable here.

File: _posts/2016-04-30-akka-basics.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+## Actors all the way
+
+Threads are an expensive resource, hence we need to use them judiciously. Blocking threads while we're waiting for some IO operation to respond is really inefficient. Actors use threads in a different way, as a result of which they're pretty lightweight ([several million actors per GB of heap memory](http://doc.akka.io/docs/akka/2.4.4/intro/what-is-akka.html))
+
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md:28]: Do you also want to say something about Event Bus here, as you are coming close to explaining about Actors - http://doc.akka.io/docs/akka/snapshot/java/event-bus.html

File: _posts/2016-04-30-akka-basics.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+> I'm sorry that I long ago coined the term "objects" for this topic because it gets many people to focus on the lesser idea. The big idea is "messaging".
+
+An actor exchanges immutable messages and keeps their own encapsulated state: every interaction needs to be done through messages.
+
+## Fault tolerance
+
+There is not a clear and shared error model in Java, and you can realise that after working in a couple of projects. The origin of that is a confusing abstraction called Exception and its subtypes Checked and Unchecked Exceptions. The only real difference is that checked expections force us to deal with the issue in the direct caller, except for that they don't hint at the strategy that we should apply.
+
+I strongly recomment [this](http://joeduffyblog.com/2016/02/07/the-error-model/) post about it. A powerful idea in the post is that bugs aren't recoverable errors. This quote is contained on that post, [original source](https://wiki.haskell.org/Error_vs._Exception):
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: ...recommend..

...in that post...

File: _posts/2016-04-30-akka-basics.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+> I was involved in the development of a library that was written in C++. One of the developers told me that the developers are divided into the ones who like exceptions and the other ones who prefer return codes. As it seem to me, the friends of return codes won. However, I got the impression that they debated the wrong point: Exceptions and return codes are equally expressive, they should however not be used to describe errors. Actually the return codes contained definitions like ARRAY_INDEX_OUT_OF_RANGE. But I wondered: How shall my function react, when it gets this return code from a subroutine? Shall it send a mail to its programmer? It could return this code to its caller in turn, but it will also not know how to cope with it. Even worse, since I cannot make assumptions about the implementation of a function, I have to expect an ARRAY_INDEX_OUT_OF_RANGE from every subroutine. My conclusion is that ARRAY_INDEX_OUT_OF_RANGE is a (programming) error. It cannot be handled or fixed at runtime, it can only be fixed by its developer. Thus there should be no according return code, but instead there should be asserts.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: 👍  nice idea

File: _posts/2016-04-30-akka-basics.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+Akka provides a great way for dealing with failures/errors based on these tenets:
+
+* Single responsibility principle. Failure management is delegated to supervisors, creating focused business actors. We agreed some time ago that lifecycle management, e.g. object creation, should be moved to factories, therefore an object is not responsible for creating itself. Resuming or restarting an object, after some failure happened, is part of that lifecycle management and Akka enforces you to move that responsibility to supervisors.
+
+* Default recovery strategies. It's important to understand which are our options after the system is under some failure condition. Is DB temporarily down? Has some input, in combination with the existing data, created an undesired state? Is that a bug? Do we need to abandon that particular request or that part of the system is messed up until some patch is applied? Those questions will determine our response to that incident, and Akka provides some [built-in strategies](http://doc.akka.io/docs/akka/2.4.4/scala/fault-tolerance.html#fault-tolerance-scala).
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: Is the DB....

File: _posts/2016-04-30-akka-basics.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+```
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: can we annotate that its scala code" ```scala, if that works

File: _posts/2016-04-30-akka-basics.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+override val supervisorStrategy =
+  OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 minute) {
+    case _: ArithmeticException      => Resume
+    case _: NullPointerException     => Restart
+    case _: IllegalArgumentException => Stop
+    case _: Exception                => Escalate
+  }
+```
+
+* Failure as first citizen. Some libraries or ecosystems hide failures through obscure APIs. Akka is usually deployed in distributed environments, therefore using unreliable resources like networks. That forces us to bring failure to the front page. Even in 'safer' environments as monoliths, failures are all around, so using toolkits like Akka is critical to create reliable and robust software.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: That forces us to bring failure to the front or surface. Dont say front page

File: _posts/2016-04-30-akka-basics.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+## Concurrency and parallelism
+
+With Akka we're not going to deal directly with threads, they're hidden under an abstraction layer. The backbone of an Akka app is the [Actor System](http://doc.akka.io/api/akka/2.0/akka/actor/ActorSystem.html):
+
+> An actor system is a hierarchical group of actors which share common configuration, e.g. dispatchers, deployments, remote capabilities and addresses. It is also the entry point for creating or looking up actors.
+
+A Dispatcher is also an Execution Context, so in the end is where the thread pool is located. Let's assume that your app uses a single dispatcher with 4 threads allocated. As you can see threads are a scarce resource, if we do blocking IO or heavy CPU work in some of the actors we'll use exclusively one of the threads, and you can see how quickly your service will run out of threads.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: Are you trying to say "so in the end it is where the thread pool is located" ?

File: _posts/2016-04-30-akka-basics.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+The solution is wrapping those operation with constructs like Scala Futures and then provide a different execution context for those tasks.
+
+## Summary
+
+Akka is a fantastic toolkit full of well suited abstractions for modern problems. In next post we'll see with code how we can coordinate and supervise actors with the constraints of our example application.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: In the next post...

Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: Is it worth coming up with a table of Akka / Actor concepts or solutions that solve high-level or low-level programming problems, nice gotcha here for interested developers.

**Comment by neomatrix369** [comment]
Minor comment: use I/O and not IO throughout the post

File: _posts/2016-04-30-akka-basics.md (hunk 3 above)
Comment (felipefzdz) [inline _posts/2016-04-30-akka-basics.md:28]: Not yet :) It will be the center of an upcoming post.

File: _posts/2016-04-30-akka-basics.md (hunk 10 above)
Comment (felipefzdz) [inline _posts/2016-04-30-akka-basics.md]: Maybe at the end of the series, I'll explain more Akka concepts in the next ones.

**Comment by neomatrix369** [comment]
Just for curiosity sake its worth running this post via word cloud generator to see the terms used terms which in turn would help find out which categories the post belongs to most, try this http://www.tripwiremagazine.com/2012/12/12-cool-word-cloud-generators.html.

File: _posts/2016-04-30-akka-basics.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+      text: Akka_20120331 by Helena Whitlock, SGU
+      href: https://www.flickr.com/photos/geologicalsurveyofsweden/8703286181
+tags:
+- akka
+- OOP
+- functional programming
+
+---
+
+As [the previous post](http://codurance.com/2016/04/28/async-systems-with-sync-clients) explained our system is based on Akka. Before going into more detail about our solution, I'd like to explain what Akka is about and why it is so awesome.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: hotlink Akka and possibly the other terms mentioned that are out of scope

File: _posts/2016-04-30-akka-basics.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+Akka is a toolkit and not a framework, you can simply use the bits that you need for your service. In this series we'll focus on the core features and we won't go through Akka Stream, Akka-Http or Akka Cluster (since they're out of scope, not because they're not fantastic).
+
+Akka provides a different abstraction to deal with concurrency, paralellism and fault-tolerance. That abstraction is called Actor Model. If you have experience with old Java ecosystem you will know how hard is [writing safe and correct multithreaded code](http://codurance.com/2015/12/13/testing-multithreaded-code-in-java/). Providing human-friendly abstractions like Actors it's a revolution similar to what Java did with memory management. New challenges derived from internet scale and cloud computing requires a reactive approach to programming. Akka aligns with that philosophy and enables you to implement [Event-driven architectures](https://www.wikiwand.com/en/Event-driven_architecture).
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: ...how hard [...link....] is.

Providing a....

File: _posts/2016-04-30-akka-basics.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+## Actors all the way
+
+Threads are an expensive resource, hence we need to use them judiciously. Blocking threads while we're waiting for some I/O operation to respond is really inefficient. Actors use threads in a different way, as a result of which they're pretty lightweight ([several million actors per GB of heap memory](http://doc.akka.io/docs/akka/2.4.4/intro/what-is-akka.html))
+
+Actors mix Object Oriented and Functional Programming principles. OOP is basically about messages as Alan Kay [stated](http://lists.squeakfoundation.org/pipermail/squeak-dev/1998-October/017019.html)
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: Actors are a good blend of Object Oriented and Functional Programming principles - if you like to say this otherwise stick to the one you have.

File: _posts/2016-04-30-akka-basics.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+> I'm sorry that I long ago coined the term "objects" for this topic because it gets many people to focus on the lesser idea. The big idea is "messaging".
+
+An actor exchanges immutable messages and keeps their own encapsulated state: every interaction needs to be done through messages.
+
+## Fault tolerance
+
+There is not a clear and shared error model in Java, and you can realise that after working in a couple of projects. The origin of that is a confusing abstraction called Exception and its subtypes Checked and Unchecked Exceptions. The only real difference is that checked expections force us to deal with the issue in the direct caller, except for that they don't hint at the strategy that we should apply.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: error handling model ?

File: _posts/2016-04-30-akka-basics.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+I strongly recommend [this](http://joeduffyblog.com/2016/02/07/the-error-model/) post about it. A powerful idea in the post is that bugs aren't recoverable errors. This quote is contained in that post, [original source](https://wiki.haskell.org/Error_vs._Exception):
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: One such powerful idea in ... - if you think it sounds better

As it seems to me....

...they debated on the wrong...

Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: surround the terms ARRAY_INDEX_OUT_OF_RANGE with `ARRAY_INDEX_OUT_OF_RANGE`

File: _posts/2016-04-30-akka-basics.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+Akka provides a great way for dealing with failures/errors based on these tenets:
+
+* Single responsibility principle. Failure management is delegated to supervisors, creating focused business actors. We agreed some time ago that lifecycle management, e.g. object creation, should be moved to factories, therefore an object is not responsible for creating itself. Resuming or restarting an object, after some failure happened, is part of that lifecycle management and Akka enforces you to move that responsibility to supervisors.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: use colons instead of the . for the bullet points.

File: _posts/2016-04-30-akka-basics.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+* Default recovery strategies. It's important to understand which are our options after the system is under some failure condition. Is the DB temporarily down? Has some input, in combination with the existing data, created an undesired state? Is that a bug? Do we need to abandon that particular request or that part of the system is messed up until some patch is applied? Those questions will determine our response to that incident, and Akka provides some [built-in strategies](http://doc.akka.io/docs/akka/2.4.4/scala/fault-tolerance.html#fault-tolerance-scala).
+
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md:48]: Reading the above paragraphs I'm thinking of design principles like `low coupling` and `high cohesion`, maybe you can say how the Actor / Reactive / Akka system is helping achieve this.

File: _posts/2016-04-30-akka-basics.md [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+override val supervisorStrategy =
+  OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 minute) {
+    case _: ArithmeticException      => Resume
+    case _: NullPointerException     => Restart
+    case _: IllegalArgumentException => Stop
+    case _: Exception                => Escalate
+  }
+```
+
+* Failure as first citizen. Some libraries or ecosystems hide failures through obscure APIs. Akka is usually deployed in distributed environments, therefore using unreliable resources like networks. That forces us to bring failure to the front. Even in 'safer' environments as monoliths, failures are all around, so using toolkits like Akka is critical to create reliable and robust software.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: first class citizen not just first citizen ;)

File: _posts/2016-04-30-akka-basics.md [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+> An actor system is a hierarchical group of actors which share common configuration, e.g. dispatchers, deployments, remote capabilities and addresses. It is also the entry point for creating or looking up actors.
+
+A Dispatcher is also an Execution Context, so in the end it is where the thread pool is located. Let's assume that your app uses a single dispatcher with 4 threads allocated. As you can see threads are a scarce resource, if we do blocking I/O or heavy CPU work in some of the actors we'll use exclusively one of the threads, and you can see how quickly your service will run out of threads.
+
+The solution is wrapping those operation with constructs like Scala Futures and then provide a different execution context for those tasks.
+
+## Summary
+
+Akka is a fantastic toolkit full of well suited abstractions for modern problems. In the next post we'll see with code how we can coordinate and supervise actors with the constraints of our example application.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: we'll see with the help of some code on how to ....

**Comment by felipefzdz** [comment]
PR updated with your comments :)

File: _posts/2016-04-30-akka-basics.md [hunk 20]
```diff
@@ -16,37 +16,37 @@ tags:
...
 
 Akka provides a great way for dealing with failures/errors based on these tenets:
 
-* Single responsibility principle. Failure management is delegated to supervisors, creating focused business actors. We agreed some time ago that lifecycle management, e.g. object creation, should be moved to factories, therefore an object is not responsible for creating itself. Resuming or restarting an object, after some failure happened, is part of that lifecycle management and Akka enforces you to move that responsibility to supervisors.
+* Single responsibility principle: failure management is delegated to supervisors, creating focused business actors. We agreed some time ago that lifecycle management, e.g. object creation, should be moved to factories, therefore an object is not responsible for creating itself. Resuming or restarting an object, after some failure happened, is part of that lifecycle management and Akka enforces you to move that responsibility to supervisors. The resulting code will be low coupled and highly cohesive.
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: loosely coupled ;)

**Comment by neomatrix369** [comment]
All looks good, nothing more to say - see if someone else has comments otherwise we can merge it.

File: _posts/2016-04-30-akka-basics.md [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+
+Akka is a toolkit and not a framework, you can simply use the bits that you need for your service. In this series we'll focus on the core features and we won't go through [Akka Streams](http://doc.akka.io/docs/akka/2.4.4/scala/stream/index.html), [Akka-Http](http://doc.akka.io/docs/akka/2.4.4/scala/http/index.html) or [Akka Cluster](http://doc.akka.io/docs/akka/2.4.4/common/cluster.html) (since they're out of scope, not because they're not fantastic).
+
+Akka provides a different abstraction to deal with concurrency, paralellism and fault-tolerance. That abstraction is called Actor Model. If you have experience with old Java ecosystem you will know how hard [writing safe and correct multithreaded code](http://codurance.com/2015/12/13/testing-multithreaded-code-in-java/) is. Providing a human-friendly abstractions like Actors it's a revolution similar to what Java did with memory management. New challenges derived from internet scale and cloud computing requires a reactive approach to programming. Akka aligns with that philosophy and enables you to implement [Event-driven architectures](https://www.wikiwand.com/en/Event-driven_architecture).
```
Comment (jukecraft) [inline _posts/2016-04-30-akka-basics.md]: `If you have experience with the old Java ecosystem`

Comment (jukecraft) [inline _posts/2016-04-30-akka-basics.md]: `you will know how hard it is to [write safe and correct multithreaded code](http://codurance.com/2015/12/13/testing-multithreaded-code-in-java/).`

Comment (jukecraft) [inline _posts/2016-04-30-akka-basics.md]: `Providing human-friendly abstractions like Actors is a revolution comparable to Java's memory management at the time of introduction.`

File: _posts/2016-04-30-akka-basics.md [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+
+## Actors all the way
+
+Threads are an expensive resource, hence we need to use them judiciously. Blocking threads while we're waiting for some I/O operation to respond is really inefficient. Actors use threads in a different way, as a result of which they're pretty lightweight ([several million actors per GB of heap memory](http://doc.akka.io/docs/akka/2.4.4/intro/what-is-akka.html))
+
+Actors are a good blend of Object Oriented and Functional Programming principles. OOP is basically about messages as Alan Kay [stated](http://lists.squeakfoundation.org/pipermail/squeak-dev/1998-October/017019.html)
```
Comment (dhatanian) [inline _posts/2016-04-30-akka-basics.md]: Maybe introduce Alan Kay? Not everybody knows him by name?

File: _posts/2016-04-30-akka-basics.md [hunk 23]
```diff
@@ -20,13 +20,13 @@ As [the previous post](http://codurance.com/2016/04/28/async-systems-with-sync-c
...
+Akka provides a different abstraction to deal with concurrency, paralellism and fault-tolerance. That abstraction is called Actor Model. If you have experience with the old Java ecosystem you will know how hard it is to [write safe and correct multithreaded code](http://codurance.com/2015/12/13/testing-multithreaded-code-in-java/). Providing human-friendly abstractions like Actors is a revolution comparable to Java's memory management at the time of introduction.. New challenges derived from internet scale and cloud computing requires a reactive approach to programming. Akka aligns with that philosophy and enables you to implement [Event-driven architectures](https://www.wikiwand.com/en/Event-driven_architecture).
```
Comment (neomatrix369) [inline _posts/2016-04-30-akka-basics.md]: .. ==> .
